packages/core/src/shared/crashMonitoring/crashMonitoring.ts

```diff
@@ -0,0 +1,626 @@
+        private readonly checkInterval: number,
+        private readonly isDevMode: boolean
+    ) {}
+
+    public start() {
+        {
+            if (this.isDevMode) {
+                getLogger().debug(`crashMonitoring: checkInterval ${this.checkInterval}`)
+            }
+
+            // IMPORTANT: Do not do an immediate check for a crash since the globalState memento can be out of sync on startup
+            // and this can lead to unexpected behavior. By waiting we assume it will be up-to-sync then.
+            this.intervalRef = globals.clock.setInterval(async () => {
+                try {
+                    await tryCheckCrash()
+                } catch (e) {
+                    emitFailure({ functionName: 'checkCrashInterval', error: e })
+                }
+            }, this.checkInterval)
+        }
+
+        // ------------ Inner Functions ------------
+
+        // nested functions do not have access to 'this', defining them here gives them access to the **value**
+        const isDevMode = this.isDevMode
+        const checkInterval = this.checkInterval
+        const state = this.state
+
+        async function tryCheckCrash() {
+            if (!(await isPrimaryChecker())) {
+                return
+            }
+
+            // Iterate all known extensions and for each check if they have crashed
+            const knownExts = state.getAllExts()
+            const runningExts: ExtInstanceHeartbeat[] = []
+            for (const ext of knownExts) {
+                if (!isStoppedHeartbeats(ext, state.now(), checkInterval)) {
+                    runningExts.push(ext)
+                    continue
+                }
+
+                // Ext is not running anymore, handle appropriately depending on why it stopped running
+                await state.handleExtNotRunning(ext, {
+                    shutdown: async () => {
+                        // Nothing to do, just log info if necessary
+                        if (isDevMode) {
+                            getLogger().debug(
+                                `crashMonitoring: SHUTDOWN: following has gracefully shutdown: pid ${ext.extHostPid} + sessionId: ${ext.sessionId}`
+                            )
+                        }
+                    },
+                    crash: async () => {
+                        // Debugger instances may incorrectly look like they crashed, so don't emit.
+                        // Example is if I hit the red square in the debug menu, it is a non-graceful shutdown. But the regular
+                        // 'x' button in the Debug IDE instance is a graceful shutdown.
+                        if (ext.isDebug) {
+                            getLogger().debug(`crashMonitoring: DEBUG instance crashed: ${JSON.stringify(ext)}`)
+                            return
+                        }
+
+                        // This is the metric to let us know the extension crashed
+                        telemetry.session_end.emit({
```

Comment:
can we force the proxied sessionId into the sessionId field? there's not really any use for the sessionId field from the current (non-proxied) session.

Reply:
sessionId is adjacent to clientId and is injected in a different part of the metric right before sending: https://github.com/aws/aws-toolkit-vscode/blob/72bd7f5097ed4d53b4746f74e08d6918103a84ca/packages/core/src/shared/telemetry/telemetryClient.ts#L109

We have to do some hacky stuff to get it to work but it is possible, though I don't think is worth the effort atm